core/templates/pages/exploration-editor-page/services/change-list.service.ts

```diff
@@ -17,246 +17,280 @@
         }
-        addChange({
-          cmd: CMD_EDIT_STATE_PROPERTY,
-          new_value: angular.copy(newValue),
-          old_value: angular.copy(oldValue),
-          property_name: backendName,
-          state_name: stateName
-        });
-      },
-      getChangeList: function() {
-        return angular.copy(explorationChangeList);
-      },
-      isExplorationLockedForEditing: function() {
-        return explorationChangeList.length > 0;
-      },
-      /**
-       * Initializes the current changeList with the one received from backend.
-       * This behavior exists only in case of an autosave.
-       *
-       * @param {object} changeList - Autosaved changeList data
-       */
-      loadAutosavedChangeList: function(changeList) {
-        explorationChangeList = changeList;
-      },
-      /**
-       * Saves a change dict that represents the renaming of a state. This
-       * is also intended to change the initial state name if necessary
-       * (that is, the latter change is implied and does not have to be
-       * recorded separately in another change dict). It is the responsibility
-       * of the caller to check that the two names are not equal.
-       *
-       * @param {string} newStateName - The new name of the state
-       * @param {string} oldStateName - The previous name of the state
-       */
-      renameState: function(newStateName, oldStateName) {
-        addChange({
-          cmd: CMD_RENAME_STATE,
-          new_state_name: newStateName,
-          old_state_name: oldStateName
-        });
-      },
-      undoLastChange: function() {
-        if (explorationChangeList.length === 0) {
-          AlertsService.addWarning('There are no changes to undo.');
-          return;
-        }
-        var lastChange = explorationChangeList.pop();
-        undoneChangeStack.push(lastChange);
-        autosaveChangeListOnChange(explorationChangeList);
-      },
-
-      get autosaveIsInProgress$(): Observable<boolean> {
-        return autosaveInProgressEventEmitter.asObservable();
+        this.autosaveInProgressEventEmitter.emit(false);
       }
-    };
+    );
   }
-]);
+
+  private addChange(changeDict) {
+    if (this.loadingMessage) {
+      return;
+    }
+    this.explorationChangeList.push(changeDict);
+    this.undoneChangeStack = [];
+    this.autosaveInProgressEventEmitter.emit(true);
+    if (this.changeListAddedTimeoutId) {
+      clearTimeout(this.changeListAddedTimeoutId);
+    }
+    this.changeListAddedTimeoutId = setTimeout(() => {
+      this.autosaveChangeListOnChange(this.explorationChangeList);
+    }, this.DEFAULT_WAIT_FOR_AUTOSAVE_MSEC);
+  }
+
+  /**
+   * Saves a change dict that represents adding a new state. It is the
+   * responsbility of the caller to check that the new state name is valid.
+   *
+   * @param {string} stateName - The name of the newly-added state
+   */
+
+  addState(stateName: string): void {
+    this.addChange({
```

Comment:
Add an object instead of a dict. (Like new Change(...))

Reply:
I couldn't understand this part can you please eloborate?